### PR TITLE
Prometheus bucket implementation

### PR DESCRIPTION
**Pull request by vsergeyev**
Hello @regel ,

This is a very draft Prometheus bucket for #92 
I have started it some time ago and due to the lack of time haven't touched for some days.
Your review is welcome. Hopefully will finish it soon.

Regards,
V.

### CODE REVIEW

File: loudml/prometheus.py [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+# https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators
+# sum (calculate sum over dimensions)
+# min (select minimum over dimensions)
+# max (select maximum over dimensions)
+# avg (calculate the average over dimensions)
+# stddev (calculate population standard deviation over dimensions)
+# stdvar (calculate population standard variance over dimensions)
+# count (count number of elements in the vector)
+# count_values (count number of elements with the same value)
+# bottomk (smallest k elements by sample value)
```
Comment (regel) [inline loudml/prometheus.py:44]: Adding bottomk and topk to AGGREGATORS gives more flexibility to users.

File: loudml/prometheus.py [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+    'min': 'min',
+    'max': 'max',
+    'sum': 'sum',
+    'deriv': None,
+    'derivative': None,
+    'integral': None,
+    'med': None,
+    'median': None,
+    'mode': None,
+    '5percentile': None,
```
Comment (regel) [inline loudml/prometheus.py:64]: 5percentile = quantile(5) if my understanding of the Prometheus operator is correct

File: loudml/prometheus.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+    """
+    Build tags predicates
+    """
+    must = []
+
+    if match_all:
+        for condition in match_all:
+            must.append('{}="{}"'.format(condition['tag'], condition['value']))
+
+    return "{" + ", ".join(must) + "}"
```
Comment (regel) [inline loudml/prometheus.py:98]: is the extra space intentional in ", " before join() ?

File: loudml/prometheus.py [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+            if aggregator:
+                query_url = "{}/api/v1/query_range?{}&step={}&query={}({}{})".format(
+                    self.url,
+                    time_pred,
+                    q["step"],
+                    aggregator,
+                    q["metric_name"],
+                    q["tags"]
+                )
+            else:
```
Comment (regel) [inline loudml/prometheus.py:178]: Shouldn't it raise UnsupportedMetric exception if the aggregation operator isn't supported?

File: loudml/prometheus.py [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+
+import numpy as np
+
+from voluptuous import (
+    Required,
+    Optional,
+    Boolean,
+)
+
+from . import (
```
Comment (regel) [inline loudml/prometheus.py:20]: from loudml import

File: loudml/prometheus.py [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+                    resp.status_code, resp.text[:200])
+
+        return results
+
+
+    def put(self, entry):
+        """
+        Store a point in database
+        """
+        raise errors.NotImplemented("Prometheus bucket: not implemented yet")
```
Comment (regel) [inline loudml/prometheus.py:206]: raise NotImplementedError

File: loudml/prometheus.py [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+    def verify_ssl(self):
+        return self.cfg.get('verify_ssl') or False
+
+    @property
+    def ssl_cert_path(self):
+        return self.cfg.get('ssl_cert_path') or ''
+
+    @property
+    def prometheus(self):
+        if self._prometheus is None:
```
Comment (regel) [inline loudml/prometheus.py:255]: `if not self._prometheus:`

If you install flake8 the "is None" comparison should trigger PEP8 warnings. The above change is a quick fix.

I use flake8 a lot. It's very useful!

File: loudml/prometheus.py [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+                verify_ssl=self.verify_ssl,
+                ssl_cert_path=self.ssl_cert_path,
+                user=self.user,
+                password=self.password
+            )
+
+        return self._prometheus
+
+    def insert_data(self, data):
+        raise errors.NotImplemented("Prometheus is a pure time-series database")
```
Comment (regel) [inline loudml/prometheus.py:275]: raise NotImplementedError

File: loudml/prometheus.py [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+        *args,
+        **kwargs
+    ):
+        """
+        Insert data
+        """
+        ts = int(make_ts(ts))
+        filtered = filter(lambda item: item[1] is not None, data.items())
+        for k, v in filtered:
+            entry = {
```
Comment (regel) [inline loudml/prometheus.py:292]: You can opt for a shorter variant:

```
            self.enqueue({
                'metric': k,
                'timestamp': ts,
                'value': v,
                'tags': tags,
            })
```

File: loudml/prometheus.py (hunk 3 above)
Comment (vsergeyev) [reply loudml/prometheus.py:98]: it's typo, thank you

File: loudml/prometheus.py (hunk 2 above)
Comment (vsergeyev) [reply loudml/prometheus.py:64]: from Prometheus docs it looks like quantile(0.05, {query expression here}), will fix in commit

File: loudml/prometheus.py (hunk 1 above)
Comment (vsergeyev) [reply loudml/prometheus.py:44]: agree

File: loudml/prometheus.py (hunk 4 above)
Comment (vsergeyev) [reply loudml/prometheus.py:178]: maybe warning in log with suggestion to submit a GitHub ticket for us should work better?

File: loudml/prometheus.py (hunk 5 above)
Comment (vsergeyev) [reply loudml/prometheus.py:20]: agree

File: loudml/prometheus.py (hunk 6 above)
Comment (vsergeyev) [reply loudml/prometheus.py:206]: agree. I also have seen a same in influx.py bucket

**Comment by vsergeyev** [comment]
Thank you @regel for your review, applied appropriate changes and added tests.
If you may help me with CI part thats would be great!

V.

**Comment by regel** [comment]
Lots of progress, awesome! Thanks

Volodymyr Github complains about conflicts so I think you have to rebase the branch and push again.

Also, you can have a look at the updated `opentsdb.py` file where you can find some useful tips: import perform_request() function can do the actual HTTP requests.

I'm looking into CI tasks, as well as bucket write, and drop functions.

**Comment by vsergeyev** [comment]
@regel please think about putting DEFAULT_REQUEST_TIMEOUT somewhere from opentsdb module, so it may be used globally

File: loudml/prometheus.py (hunk 4 above)
Comment (regel) [reply loudml/prometheus.py]: Yes good idea

File: loudml/prometheus.py [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+
+    def build_url_params(self, q):
+        """
+        Forms a query URL params from bits.
+        TODO: add better aggregator functions handling
+        """
+        params = {
+            'start': int(make_ts(q['start'])),
+            'end': int(make_ts(q['end'])),
+            'step': q["step"]
```
Comment (regel) [inline loudml/prometheus.py:154]: The API docs mention a timeout. Should we try to set it this field in params?

```
query=<string>: Prometheus expression query string.
time=<rfc3339 | unix_timestamp>: Evaluation timestamp. Optional.
timeout=<duration>: Evaluation timeout. Optional. Defaults to and is capped by the value of the -query.timeout flag.
```

timeout: DEFAULT_REQUEST_TIMEOUT

**Comment by regel** [comment]
I'm looking at the options we have for a minimal put() implementation and it's tricky:(. Unfortunately, there is no easy push model except the GraphiteBridge in prometheus_client library.

**Comment by vsergeyev** [comment]
Have updated pull request based on recent comments.
Also moved "DEFAULT_REQUEST_TIMEOUT" into loudml.requests in one commit.

**Comment by regel** [comment]
Hi Servey, thanks so much! I will look into bucket write capabilities probably over the week-end.

In the meantime, can you solve the git rebase conflict and push this PR again?

Here is how (simple setup.py file conflict to resolve manually):

```
[regel@mac loudml]$ git rebase origin/develop
First, rewinding head to replay your work on top of it...
Applying: Prometheus datasource/bucket development started
Using index info to reconstruct a base tree...
M	setup.py
Falling back to patching base and 3-way merge...
Auto-merging setup.py
CONFLICT (content): Merge conflict in setup.py
Failed to merge in the changes.
Patch failed at 0001 Prometheus datasource/bucket development started
The copy of the patch that failed is found in:
   /home/regel/src/github.com/regel/loudml/.git/rebase-apply/patch

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".

[regel@mac loudml]$ vim setup.py 
[regel@mac loudml]$ git add setup.py 
[regel@mac loudml]$ git rebase --continue
Applying: Prometheus datasource/bucket development started
Applying: Prometheus bucket: finished client's query() and bucket's get_times_data()
Applying: Added tests
Applying: Prometheus bucket added to entry point
Applying: Prometheus bucket: applied review comments; added quantiles, bottomk, topk functions; added query URL test
Applying: Updated prometheus bucket based on requests refactoring
Applying: Refactoring: moved DEFAULT_REQUEST_TIMEOUT to loudml.requests
Applying: Added timeout to Prometheus query
```

**Comment by vsergeyev** [comment]
Hello @regel 

Hope you doing well these days.

I have run a rebase and see new commits with same names added.
Maybe it is worth to squash all of this into 1 commit when you will do a merge?

V.

**Comment by regel** [comment]
Sure, good idea. I can squash and merge. Thank you V!

**Pull request merged**